src/dnaio/_core.pyx

```diff
@@ -258,122 +266,133 @@ def fastq_iter(file, sequence_class, Py_ssize_t buffer_size):
+            # newline. Because in case of the end of buffer, name_end will be
+            # bufend - 1. So sequence_start will be bufend. The search
+            # length will be bufend - bufend == 0 and memchr will return NULL.
+            name_end = name_end_ptr - c_buf
+            sequence_start = name_end + 1
+            sequence_end_ptr = <char *>memchr(c_buf + sequence_start, 10, <size_t>(bufend - sequence_start))
+            if sequence_end_ptr == NULL:
                 break
-            endskip = 1 if c_buf[pos-1] == b'\r' else 0
-            sequence = c_buf[sequence_start:pos-endskip].decode('latin-1')
-            sequence_length = pos - endskip - sequence_start
-            pos += 1
-
-            # Parse second header (line 2)
-            second_header_start = pos
-            if pos == bufend:
+            sequence_end = sequence_end_ptr - c_buf
+            second_header_start = sequence_end + 1
+            second_header_end_ptr = <char *>memchr(c_buf + second_header_start, 10, <size_t>(bufend - second_header_start))
+            if second_header_end_ptr == NULL:
                 break
-            if c_buf[pos] != b'+':
+            second_header_end = second_header_end_ptr - c_buf
+            qualities_start = second_header_end + 1
+            qualities_end_ptr = <char *>memchr(c_buf + qualities_start, 10, <size_t>(bufend - qualities_start))
+            if qualities_end_ptr == NULL:
+                break
+            qualities_end = qualities_end_ptr - c_buf
+
+            if c_buf[record_start] != b'@':
+                raise FastqFormatError("Line expected to "
+                    "start with '@', but found {!r}".format(chr(c_buf[record_start])),
+                    line=n_records * 4)
+            if c_buf[second_header_start] != b'+':
                 raise FastqFormatError("Line expected to "
-                    "start with '+', but found {!r}".format(chr(c_buf[pos])),
+                    "start with '+', but found {!r}".format(chr(c_buf[second_header_start])),
                     line=n_records * 4 + 2)
-            pos += 1  # skip over the '+'
-            while pos < bufend and c_buf[pos] != b'\n':
-                pos += 1
-            if pos == bufend:
-                break
-            endskip = 1 if c_buf[pos-1] == b'\r' else 0
-            second_header_length = pos - endskip - second_header_start - 1
-            if second_header_length == 0:
-                second_header = False
-            else:
+
+            name_start = record_start + 1  # Skip @
+            second_header_start += 1  # Skip +
+            name_length = name_end - name_start
+            sequence_length = sequence_end - sequence_start
+            second_header_length = second_header_end - second_header_start
+            qualities_length = qualities_end - qualities_start
+
+            # Check for \r\n line-endings and compensate
+            if c_buf[name_end - 1] == b'\r':
+                name_length -= 1
```

Comment:
See #26 . The actual parsing of the FASTQ is not where most of the compute power is now. Since the addition of memchr it is actually in creating the string objects.
So I am inclined to keep the "human-readable" version here.
We can benchmark it later indeed, in a separate PR, but I doubt the performance will be noticable.

What I do think matters more here is the actual accessing of the contents of `c_buf`. We have just read `c_buf` from `record_start` to `qualities_end`, but then we jump back again to the positions `name_end`, `sequence_end`, `second_header_end` and `qualities_end`. 
Next we iterate from record_start to sequence_end again when creating the string objects.
So now we traverse the same array of memory three times. Once for finding the newlines, once for checking the line endings and once for the creation of the python objects. I think the newline and line end checking can be combined. But I do wonder how much performance it will add.

Reply:
Sure, keep that version, and as I said, I expect the difference to be marginal at best.

Regrouping the statements sounds like an easy thing to do to improve cache locality a little bit. My guess is it won’t help when the reads are short since everything should be in the first-level cache anyway.